python/langsmith/run_trees.py

```diff
@@ -233,13 +233,53 @@ class Config:
     @root_validator(pre=True)
     def infer_defaults(cls, values: dict) -> dict:
         """Assign name to the run."""
+        # Try to get name from serialized data (legacy LangChain support)
         if values.get("name") is None and values.get("serialized") is not None:
-            if "name" in values["serialized"]:
-                values["name"] = values["serialized"]["name"]
-            elif "id" in values["serialized"]:
-                values["name"] = values["serialized"]["id"][-1]
+            serialized = values["serialized"]
+            values["name"] = serialized.get("name") or (
+                serialized["id"][-1] if "id" in serialized else None
+            )
+
+        # If still no name, try to infer from caller
         if values.get("name") is None:
-            values["name"] = "Unnamed"
+            name = None
+            try:
+                import inspect
+                import os
+
+                # Look up the stack, skipping internal frames
+                for frame_info in inspect.stack()[1:8]:
+                    filename = frame_info.filename
+                    norm_path = filename.replace("\\", "/")
+
+                    # Skip pydantic and langsmith internal modules
+                    if "/pydantic/" in norm_path:
+                        continue
+                    if os.path.basename(filename) in {
+                        "run_trees.py",
+                        "run_helpers.py",
+                        "client.py",
+                        "_context.py",
+                    }:
+                        continue
+
+                    # Try to get class name from self or cls
+                    frame_locals = frame_info.frame.f_locals
+                    if "self" in frame_locals:
+                        name = frame_locals["self"].__class__.__name__
+                        break
+                    elif "cls" in frame_locals:
+                        name = frame_locals["cls"].__name__
```

Comment:
I don't know enough about Python to know if this is a good idea

Is there no easier fallback?